x-pack/plugins/security_solution/scripts/endpoint/sentinelone_host/common.ts

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { ToolingLog } from '@kbn/tooling-log';
+import type { AxiosRequestConfig } from 'axios';
+import axios from 'axios';
+import { createToolingLogger } from '../../../common/endpoint/data_loaders/utils';
+import type {
+  S1SitesListApiResponse,
+  S1AgentPackage,
+  S1AgentPackageListApiResponse,
+} from './types';
+import { catchAxiosErrorFormatAndThrow } from '../common/format_axios_error';
+import type { HostVm } from '../common/types';
+
+interface S1ClientOptions {
+  /** The base URL for SentinelOne */
+  url: string;
+  /** The API token that should be used to communicate with SentinelOne */
+  apiToken: string;
+  log?: ToolingLog;
+}
+
+export class S1Client {
+  protected readonly API_SITES_PATH = '/web/api/v2.1/sites';
+  protected readonly API_AGENT_PACKAGES_PATH = '/web/api/v2.1/update/agent/packages';
+
+  protected readonly log: ToolingLog;
+  protected readonly setup: Promise<{
+    siteToken: string;
+    siteId: string;
+  }>;
+
+  constructor(private readonly options: S1ClientOptions) {
+    this.log = options.log ?? createToolingLogger();
+
+    this.setup = this.request<S1SitesListApiResponse>({
+      url: this.API_SITES_PATH,
+      params: {
+        isDefault: true,
+      },
+    }).then((response) => {
+      this.log.verbose(response);
+
+      const site = response.data.sites[0];
+
+      if (!site) {
+        throw new Error(
+          `Unable to retrieve SentinelOne Site information. No default sites returned`
+        );
+      }
+
+      return {
+        siteToken: site.registrationToken,
+        siteId: site.id,
+      };
+    });
+  }
+
+  protected async request<T = unknown>({
+    url = '',
+    params = {},
+    ...options
+  }: AxiosRequestConfig): Promise<T> {
+    const apiFullUrl = this.buildUrl(url);
+
+    const requestOptions: AxiosRequestConfig = {
+      ...options,
+      url: apiFullUrl,
+      params: {
+        APIToken: this.options.apiToken,
+        ...params,
+      },
+    };
+
+    this.log.debug(`Request: `, requestOptions);
+
+    return axios
+      .request<T>(requestOptions)
+      .then((response) => {
+        this.log.verbose(`Response: `, response);
+        return response.data;
+      })
+      .catch(catchAxiosErrorFormatAndThrow);
+  }
+
+  public buildUrl(path: string): string {
+    const uri = new URL(this.options.url);
+    uri.pathname = path;
+    return uri.toString();
+  }
+
+  public async getSiteToken(): Promise<string> {
+    return this.setup.then(({ siteToken }) => siteToken);
```

Comment:
Not sure about type here, it looks like it's returning a resolved siteToken, but the interface says Promise<string> - is this correct?

Reply:
@tomsonpl - `this.setup()` is a Promise that resolves to an object containing `siteToken` (a `string`). the `.then()` takes that property from the resolved object and returns its value  So the return type matches - it returns a promise that resolves to a `string`